keps/sig-node/3386-kubelet-evented-pleg/README.md

```diff
@@ -0,0 +1,316 @@
+
+
+### User Stories
+
+- As a cluster administrator I want to enable `Evented PLEG` feature of the kubelet for better performance.
+   - They will have to set a kubelet argument as mentioned [here](###New-Kubelet-Parameter).
+
+
+### Risks and Mitigations
+
+PLEG is very core to the container status handling in the kubelet. Hence any miscalculation there would result in unpredictible behaviour not just for the node but for an entire cluster.
+
+To reduce the risk of regression, this feature initially will be available only as an opt-in. As discussed in the section [New Kubelet Parameter](###New-Kubelet-Parameter) unless this parameter is exclusively set, the kubelet PLEG will continue to function using frequent relisting as it's doing it today. If after enabling this feature, users notice any discrepancy in container statuses they can restart the kubelet either by omitting that option or setting it to `false` explicitly.
+
+## Design Details
+
+### New Kubelet Parameter
+A new kubelet parameter, `EventedPleg`, will be introduced that can be set via [a config file](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/) or as command line argument (i.e. `--evented-pleg=true`).
+
+If the value of this parameter is set to `true` only then the kubelet will utilize CRI events otherwise the PLEG will continue to function using frequent relisting.
+
+### Runtime Service Changes
+
+A new RPC will be introduced in the [CRI Runtime Service](https://github.com/kubernetes/kubernetes/blob/6efd6582df2011f1ec8c146ef711b3348ae07d60/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L34),
+```protobuf=
+    // GetContainerEvents gets container events from the CRI runtime
+    rpc  GetContainerEvents(GetEventsRequest) returns (stream ContainerEventResponse) {}
+```
+
+```protobuf=
+message ContainerEventResponse {
+    // ID of the container
+    string container_id = 1;
+
+    // Type of the container event
+    ContainerEventType container_event_type = 2;
+
+    // Creation timestamp of this event
+    int64 created_at = 3;
+
+    // ID of the sandbox container
+    string sandbox_id = 4;
+}
+```
+
+```protobuf=
+enum ContainerEventType {
+    // Container created
+    CONTAINER_CREATED_EVENT = 0;
+
+    // Container started
+    CONTAINER_STARTED_EVENT = 1;
+
+    // Container stopped
+    CONTAINER_STOPPED_EVENT = 2;
+
+    // Container deleted
+    CONTAINER_DELETED_EVENT = 3;
+}
+```
+### Image Service Changes
```

Comment:
I think we should narrow down the scope for the first iteration to just the container lifecycle events.